src/main/java/no/unit/nva/database/DatabaseServiceImpl.java

```diff
@@ -57,61 +65,126 @@ public DatabaseServiceImpl(DynamoDBMapper mapper) {
     }
 
     @Override
-    public Optional<UserDto> getUser(UserDto queryObject) throws InvalidUserInternalException {
-        logger.debug(
-            GET_USER_DEBUG_MESSAGE + convertToStringOrWriteErrorMessage(queryObject));
-        DynamoDBQueryExpression<UserDb> searchUserRequest = createGetQuery(queryObject.toUserDb());
-        List<UserDb> userSearchResult = mapper.query(UserDb.class, searchUserRequest);
-
-        UserDto user = userSearchResult
-            .stream()
-            .map(attempt(UserDto::fromUserDb))
-            .map(attempt -> attempt.orElseThrow(this::unexpectedException))
-            .collect(SingletonCollector.collectOrElse(null));
-
-        return Optional.ofNullable(user);
+    public UserDto getUser(UserDto queryObject) throws InvalidEntryInternalException, NotFoundException {
+        Optional<UserDto> resultOpt = getUserAsOptional(queryObject);
+        return resultOpt.orElseThrow(() -> new NotFoundException(USER_NOT_FOUND_MESSAGE + queryObject.getUsername()));
     }
 
     @Override
-    public void addUser(UserDto user) throws InvalidUserInternalException, ConflictException {
+    public void addUser(UserDto user) throws InvalidEntryInternalException, ConflictException, InvalidInputException {
         logger.debug(ADD_USER_DEBUG_MESSAGE + convertToStringOrWriteErrorMessage(user));
-        if (userAlreadyExists(user)) {
-            throw new ConflictException(USER_ALREADY_EXISTS_ERROR_MESSAGE + user.getUsername());
-        }
+
+        validate(user);
+        checkUserDoesNotAlreadyExist(user);
         mapper.save(user.toUserDb());
     }
 
     @Override
-    public void addRole(RoleDto roleDto) throws InvalidInputRoleException {
+    public void addRole(RoleDto roleDto) throws ConflictException, InvalidInputException,
+                                                InvalidEntryInternalException {
+
         logger.debug(ADD_ROLE_DEBUG_MESSAGE + convertToStringOrWriteErrorMessage(roleDto));
-        Try.of(roleDto)
-            .forEach(role -> mapper.save(roleDto.toRoleDb()))
-            .orElseThrow(failure -> new InvalidInputRoleException(INVALID_ROLE_ERROR));
+
+        validate(roleDto);
+        checkRoleDoesNotExist(roleDto);
+        mapper.save(roleDto.toRoleDb());
+    }
+
+    @Override
+    public void updateUser(UserDto queryObject)
```

Comment:
only functionality tested implemented is updating existing user. More tests will be added to for this functionality later